### DISCUSSION TITLE
Scroling document by text selection

### DISCUSSION THREAD

**Discussion by KrzysztofDusko** [AvaloniaUI/AvaloniaEdit#388] [Q&A]
I works with long SQL codes. Its common to select very long SQL and run it. In AvaloniaEdit selecting text by mouse above first visible line is not possible.. for some reason document is not scrolled automaticly.
Using Shift + arrow works fine- document is automaticly scrolled.
![image](https://github.com/AvaloniaUI/AvaloniaEdit/assets/69449360/be43ddca-bc13-4fd7-9625-9ffb1f57ec9e)
To reproduce try to select lines from 81 to 1 by mouse... 
Simmilar issus is with selection is oposite direction from line 1 to 81 its not possible (by left mouse button + mouse gesture), document scrolls but only for few lines.. 

Is it possble to change this behavior ?

**Comment by danipen** [comment]
@Gillibald do you think this could be fixed? AFAIK the textbox is able tondo that so it would be applying the same here I guess...

**Comment by Gillibald** [comment]
I am not so familiar with the code but it sounds like this should be possible

**Comment by danipen** [comment]
Can you please refer to the Avalonia code that fires the scrolling when the mouse is reaching the edges for reference?

**Comment by Gillibald** [comment]
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Presenters/TextPresenter.cs#L501

**Comment by KrzysztofDusko** [comment]
i make some observations about this situation maybe this will be helpful.

scroling (with mouse selection) is possible only for about 200 lines and stops..  becouse GetVisualLineFromVisualTop returns  null at some point...https://github.com/AvaloniaUI/AvaloniaEdit/blob/be8fc934ba87692bc3d4a2e5360ae595c115b5f8/src/AvaloniaEdit/Editing/SelectionMouseHandler.cs#L585-L593
https://github.com/AvaloniaUI/AvaloniaEdit/blob/be8fc934ba87692bc3d4a2e5360ae595c115b5f8/src/AvaloniaEdit/Rendering/TextView.cs#L1648-L1661
when offset is -1 SetCaretOffsetToMousePosition fails.. 
https://github.com/AvaloniaUI/AvaloniaEdit/blob/be8fc934ba87692bc3d4a2e5360ae595c115b5f8/src/AvaloniaEdit/Editing/SelectionMouseHandler.cs#L675

i tried to fix GetVisualLineFromVisualTop by adding

```
if (pos.Y >= textView.VisualLines[^1].VisualTop + textView.VisualLines[^1].Height)
    pos = pos.WithY(textView.VisualLines[^1].VisualTop + textView.VisualLines[^1].Height - ExtensionMethods.Epsilon);
```
to GetOffsetFromMousePosition becouse the is already simillar code 
```
if (pos.Y >= textView.DocumentHeight)
    pos = pos.WithY(textView.DocumentHeight - ExtensionMethods.Epsilon);
```
but this only cousing that scroll works for only few more lines...

code that "works" (and probably breaks some other scenario..) is : 

instead
https://github.com/AvaloniaUI/AvaloniaEdit/blob/be8fc934ba87692bc3d4a2e5360ae595c115b5f8/src/AvaloniaEdit/Editing/SelectionMouseHandler.cs#L653-L678


```c#
       private void SetCaretOffsetToMousePosition(PointerEventArgs e, ISegment allowedSegment = null)
       {
           int visualColumn;
           bool isAtEndOfLine;
           int offset;
           if (_mode == SelectionMode.Rectangular)
           {
               offset = GetOffsetFromMousePositionFirstTextLineOnly(e.GetPosition(TextArea.TextView), out visualColumn);
               isAtEndOfLine = true;
           }
           else
           {
               offset = GetOffsetFromMousePosition(e, out visualColumn, out isAtEndOfLine);
           }

           if (allowedSegment != null)
           {
               offset = offset.CoerceValue(allowedSegment.Offset, allowedSegment.EndOffset);
           }

           if (offset == -1)
           {
               var tmpLoc = TextArea.Caret.Position.Location;
               int currentOffset = TextArea.Document.GetOffset(tmpLoc);
               int addToOffset = TextArea.Document.Lines[tmpLoc.Line].Length;
               if (tmpLoc.Line < TextArea.Document.LineCount - 1)
               {
                   addToOffset += TextArea.Document.Lines[tmpLoc.Line +1].Length;
               }
               if (currentOffset + addToOffset < TextArea.Document.TextLength)
               {
                   offset = currentOffset + addToOffset;
               }
           }

           if (offset >= 0)
           {
               var loc = TextArea.Document.GetLocation(offset);
               var newPos = new TextViewPosition(loc, visualColumn) { IsAtEndOfLine = isAtEndOfLine };
               Debug.WriteLine("old pos" + TextArea.Caret.Position);
               Debug.WriteLine("new pos" + newPos);
               if (newPos == TextArea.Caret.Position && _mode == SelectionMode.Normal)
               {
                   double actualY = e.GetPosition(TextArea.TextView).Y;
                   if (actualY > TextArea.TextView.Bounds.Height && loc.Line < TextArea.Document.LineCount) // fix scroll down
                   {
                       newPos = new TextViewPosition(loc.Line + 1, 0, visualColumn) { IsAtEndOfLine = isAtEndOfLine };
                   }
                   else if (actualY < 0 && loc.Line > 0) // fix scroll up
                   {
                       newPos = new TextViewPosition(loc.Line - 1, 0, visualColumn) { IsAtEndOfLine = isAtEndOfLine };
                   }
               }

               TextArea.Caret.Position = newPos;
               TextArea.Caret.DesiredXPos = double.NaN;
           }
           //else
           //{
           //    var line = TextArea.Caret.Position.Line + 1;
           //    var column = 0;
           //    TextArea.Caret.Position = new TextViewPosition(new TextLocation(line, column), visualColumn) { IsAtEndOfLine = isAtEndOfLine };
           //    TextArea.Caret.DesiredXPos = double.NaN;
           //}
       }

```

**Comment by danipen** [comment]
We should also check the [original implementation (AvalonEdit) ](https://github.com/icsharpcode/AvalonEdit)... to see if all is ported fine.

**Comment by KrzysztofDusko** [comment]
In avalon edis issue is not present.

**Comment by KrzysztofDusko** [comment]
in WPF implemetation when selection scroling starts some part from above line is visible.... That imples one more visual line is avaiable
![image](https://github.com/AvaloniaUI/AvaloniaEdit/assets/69449360/5836a9ad-e726-417e-a054-200cf5df194e)

same when scroling down 
![image](https://github.com/AvaloniaUI/AvaloniaEdit/assets/69449360/1469a315-62b1-4c28-a6ea-a2dbe2f9913b)

without selecion starts and eds looks "normal" (no fraction of line is visible)
![image](https://github.com/AvaloniaUI/AvaloniaEdit/assets/69449360/320f1bc0-eb65-486e-b031-832eb0370119)

in Avalonia implementation scrolling up do not trigger because no visual line number 65. on screen
![image](https://github.com/AvaloniaUI/AvaloniaEdit/assets/69449360/8e4786b7-41c2-464e-b3d3-d1eb75660c28)

Question is.. can we use the same idea as AvalonEdit used ?

Edit.. WOW this is very simple.. 
TextArea.Caret.BringCaretToView(5.0);
vs
textArea.Caret.BringCaretToView(0);

is ExtendSelectionToMouse.. 


https://github.com/AvaloniaUI/AvaloniaEdit/blob/be8fc934ba87692bc3d4a2e5360ae595c115b5f8/src/AvaloniaEdit/Editing/SelectionMouseHandler.cs#L705

https://github.com/icsharpcode/AvalonEdit/blob/30cad99ce905412ed5f5e847e3c00e72e69aee77/ICSharpCode.AvalonEdit/Editing/SelectionMouseHandler.cs#L653

PR: 
https://github.com/AvaloniaUI/AvaloniaEdit/pull/390

**Comment by danipen** [comment]
Thanks for the PR. In fact I remember removing that extra space to avoid scroll jumps... See related PR. https://github.com/AvaloniaUI/AvaloniaEdit/pull/254

I'm guessing if we can add some logic to make the scrolling only when we're dragging...

**Comment by KrzysztofDusko** [comment]
well i changed TextArea.Caret.BringCaretToView(0) -> TextArea.Caret.BringCaretToView(5) only in ExtendSelectionToMouse method.
But if still there is causing problem with "sclroll jums" (i cant't reprocduce this issue). Mayby some this "hack" will by ok for both worlds ?
TextArea.Caret.BringCaretToView(**0.05**); (0.05 still trigger scrolling and should not have visual effect)

**Comment by KrzysztofDusko** [comment]
https://github.com/AvaloniaUI/AvaloniaEdit/pull/390
